File: Leetcode_oj/Binary Tree Preorder Traversal_stack/Binary Tree Preorder Traversal.cpp
// Binary Tree Preorder Traversal - do with stack
#include<iostream>
#include<stack>

using namespace std;

// Definition for binary tree
struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };
 
vector<int> preorderTraversal(TreeNode *root) {

    vector<int> res;
    if(root==NULL)
		return res;
	
    TreeNode* node=root;
    res.push_back(node->val); 

	stack<TreeNode*> stackNodes;

    if(node->right) // 先右子树 
        stackNodes.push(node->right);
    if(node->left) // 后左子树 
        stackNodes.push(node->left);

    while(!stackNodes.empty()){

        node = stackNodes.top();
        stackNodes.pop();

        res.push_back(node->val);

        if(node->right)
            stackNodes.push(node->right);
        if(node->left)
            stackNodes.push(node->left);
    } 
    return res;
}

File: Leetcode_oj/Binary Tree Preorder Traversal.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        if root==None :
            return []
        else:
            left_order=self.preorderTraversal(root.left)
            right_order=self.preorderTraversal(root.right)
            inorder=[root.val]
            inorder=inorder+left_order+right_order
            return inorder

if __name__=='__main__':
    
    s=Solution()
    
    x=TreeNode(1)
    y=TreeNode(2)
    z=TreeNode(3)
    x.right=y
    y.left=z
    
    print(s.preorderTraversal(x))
    
    a=[1,2,3]
    b=[4,5,6]
    print(a+b)